### ISSUE TITLE
How to pass docker socket as volume? 

### ISSUE THREAD

**Issue by megastef** [portainer/templates#37]
Hi, 

the volume entry in the template is a simple array. How could i pass 
```-v /var/run/docker.sock:/var/run/docker.sock```? 
 
I'm working on a template entry and access to docker socket is mandatory for it, see: https://gist.github.com/megastef/0a97da2cecd619e643edf2a1307c6fb4

Ahh, one more question - is there an option to run a global Swarm service via template?

Thx
Stefan

**Comment by deviantony** [comment]
Hi @megastef,

Our template system does not support bind mounts currently. Only volumes. When you declare a volume in a template definition, Docker will automatically create a volume and map it to the path specified in the template.

So in your case, you won't be able to achieve what you want... yet.

And, no, templates are container only for now and Swarm services are not supported... yet.

**Comment by otisg** [comment]
@deviantony I think I saw demos/videos of Portainer and Swarm the other day.  Is access to docker.sock possible now, as @megastef asked above?

**Comment by deviantony** [comment]
@otisg still not possible **inside** a template definition. This is not related with #42 (ability to specify a command for a template, e.g. `echo test` for example).

Did you saw something related to this issue in the demo?

**Comment by deviantony** [comment]
Hi @otisg @megastef 

Once https://github.com/portainer/portainer/issues/587 is implemented, you will be able to define the path to the Docker socket inside the `volumes` field such as `volumes: ["/var/run/docker.sock"]` and you'll be able to map any volume/host path onto it from the UI.

**Comment by deviantony** [comment]
This is now integrated in Portainer 1.12.

**Comment by megastef** [comment]
@deviantony Great, thank you! I will make a PR with the template soon.

**Comment by deviantony** [comment]
Reopened, we need to find a way to easily bind the Docker socket in the templates.

Detailed problem in #52

**Comment by megastef** [comment]
Hi, could I suggest a format for volumes? 

```
"volumes": [
  {
    "source": "/var/run/docker.sock",
    "target": "/var/run/docker.sock",
    "type": "bind"
  } 
 ]
```
The settings specified would be the defaults in the Portainer UI.
It should not be a problem in JavaScript to have mixed format with old and new format (as you could check for String or Object type and required properties). 

BTW, you could get some inspiration in the Rancher Catalog for app templates, UI settings etc. 
https://github.com/rancher/community-catalog/tree/master/templates/sematext

Basically there is one file [mapping UI entries to variable names](https://github.com/rancher/community-catalog/blob/master/templates/sematext/0/rancher-compose.yml), and kind of [docker-compose file using the variables, entered by the user](https://github.com/rancher/community-catalog/blob/master/templates/sematext/0/rancher-compose.yml).

**Comment by megastef** [comment]
Any progress here? 
In my PR https://github.com/portainer/templates/pull/52 
the Docker socket was working. Users had to choose the right binding manually, but it worked. 
Why don't you simply check for the Docker socket path to set it right. 
As we see there are more tools with the need of this feature.

**Comment by deviantony** [comment]
@megastef see https://github.com/portainer/portainer/issues/777 for the evolution on this

**Comment by deviantony** [comment]
https://github.com/portainer/portainer/issues/777 has been implemented ! Docs available via http://portainer.readthedocs.io/en/stable/templates.html#volumes
